rfd/0096-securing-docker-images.md

```diff
@@ -0,0 +1,323 @@
+base images.
+
+The `cosign` tool [integrates well with GHA](https://github.blog/2021-12-06-safeguard-container-signing-capability-actions/), 
+and is even included in the template "how to publish a docker image" example 
+workflow. It also provides "keyless" signing via OIDC, meaning that our build
+process can use its GitHub identity to sign the image, obviating the need for
+extra keys for us to manage.
+
+More information keyless signing:
+ * https://docs.sigstore.dev/cosign/sign/#keyless-signing
+ * https://docs.google.com/document/d/1461lQUoVqbhCve7PuKNf-2_NfpjrzBG5tFohuMVTuK4/edit
+ * https://www.appvia.io/blog/tutorial-keyless-sign-and-verify-your-container-images/#oidc-flow-with-interaction-free-with-github-actions
+
+### Build-time scanning
+
+All Teleport images shall be scanned with [trivy](https://github.com/aquasecurity/trivy-action#using-trivy-with-github-code-scanning) 
+at build time, and if any high-risk vulnerabilities are found the build will be stopped.
+
+### Image build process summary
+
+```mermaid
+graph TD
+    pull_base[Pull distroless<br/>base image]
+    signature_valid?{is cosign<br/>signature valid?}
+    deb[/Teleport Debian Package<br/>from CI/]
+    df[/Dockerfile from repo/]
+    base[/Distroless base image/]
+    candidate_image[/Candidate Telport image/]
+    push_internal[Push Candidate Image<br/>to internal ECR Registry]
+    trivy_scan[Scan Candidate Image<br/>with trivy]
+    build[$ docker build ...] 
+    upload_results[Upload scan results<br/>to GitHub Code Scanning]
+    bad_scan?{Does scan<br/>detect high-risk<br/>vulnerabilities?}
+    fail_build[Fail the build]
+    sign_image[$ cosign ...]
+    promote_build[/Release engineer<br/>promotes build/]
+    promote_image[Candidate image<br/>becomes release image<br/>in Public ECR]
+    build_ok?{Does all go well<br/>with the rest of<br/>the build?}
+    
+    pull_base --> signature_valid?
+    signature_valid? -- yes --> base
+    signature_valid? -- no --> fail_build
+    deb --> build
+    df --> build
+    base --> build
+    build --> sign_image --> candidate_image
+    candidate_image --> push_internal
+    push_internal --> trivy_scan
+    trivy_scan --> upload_results 
+    upload_results --> bad_scan?
+    bad_scan? -- no --> build_ok?
+    bad_scan? -- yes --> fail_build
+    promote_build --> promote_image
+    build_ok? -- yes --> promote_build
+    build_ok? -- no --> fail_build
+```    
+
+### Debug Images
+
+Troubleshooting a distroless image is hard, as there are no tools baked into
+the image to aid in debugging a deployment.
+
+In addtion to the main distroless image sets, the `distroless` team also supplies 
+a `debug`-tagged image that includes `busybox` (that is, a basic shell and some
+utilities). In order to provide tooling for troubleshooting a Teleport installation,
+we will create a parallel `teleport-debug` image, based on a distroless `debug` image.
```

Comment:
Have we considered going the opposite route, where we add a `minimal` tag that is designed to have only a few packages, updated regularly? This may result in less breaking changes for our customers.